### PR TITLE
Amortize stream costs

### LINKED ISSUE #471: Can Middleware modify Response In OnCompleted callback?
Looking at [FrameOfT#98](https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs#L98).

It doesn't seem like there's anything to prevent Middleware from modifying the response in the `OnCompleted` callback.

Does `OnCompleted` mean the application response has been received by the server, and so Middleware can add extra stuff to the response stream (diagnostics for instance) 
 or does it mean the response has been fully received and has been sent (or is being sent) to the client, so Middleware can clean up any resources attached to this request?

### PR DESCRIPTION
**Pull request by benaadams**
Reuse streams over connection lifetime.

Reduction in 12,000 objects per 4k requests; and 640000 bytes

Also resolves #471 "Can Middleware modify Response In OnCompleted callback" /cc @cesarbs

### CODE REVIEW

**Comment by benaadams** [comment]
Byte allocs

before

after

Allocations on `(!_requestProcessingStopping)` line are for:
`var messageBody = MessageBody.For(HttpVersion, _requestHeaders, this);`

**Comment by benaadams** [comment]
Will add a flag for this...

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs [hunk 1]
```diff
@@ -64,11 +64,15 @@ public class Frame<TContext> : Frame
                     {
                         var messageBody = MessageBody.For(HttpVersion, _requestHeaders, this);
                         _keepAlive = messageBody.RequestKeepAlive;
-                        _requestBody = new FrameRequestStream(messageBody);
-                        RequestBody = _requestBody;
-                        _responseBody = new FrameResponseStream(this);
-                        ResponseBody = _responseBody;
-                        DuplexStream = new FrameDuplexStream(RequestBody, ResponseBody);
+                        if (!ReuseStreams)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: nit: space before if block

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs [hunk 2]
```diff
@@ -64,11 +64,15 @@ public class Frame<TContext> : Frame
...
+                        {
+                            _requestBody = new FrameRequestStream();
+                            _responseBody = new FrameResponseStream(this);
+                            _duplexStream = new FrameDuplexStream(_requestBody, _responseBody);
+                        }
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: nit: new line after if block

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs [hunk 3]
```diff
@@ -93,6 +97,9 @@ public class Frame<TContext> : Frame
                                 await FireOnStarting();
                             }
 
+                            _requestBody.PauseAcceptingReads();
+                            _responseBody.PauseAcceptingWrites();
+
                             if (_onCompleted != null)
                             {
                                 await FireOnCompleted();
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: Could this throw?

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs [hunk 4]
```diff
@@ -135,7 +160,7 @@ private void ValidateState()
             {
                 case StreamState.Open:
                     return;
-                case StreamState.Disposed:
+                case StreamState.Closed:
                     throw new ObjectDisposedException(nameof(FrameRequestStream));
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: A few thoughts here:

1) `ObjectDisposedException` doesn't seem like the right exception to throw here since the stream can switch between the `Open` and `Closed` states multiple times.
2) Should we really throw? How about just no-oping when someone attempts to operate on a closed stream?

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameResponseStream.cs [hunk 5]
```diff
@@ -100,7 +125,7 @@ private void ValidateState()
             {
                 case StreamState.Open:
                     return;
-                case StreamState.Disposed:
+                case StreamState.Closed:
                     throw new ObjectDisposedException(nameof(FrameResponseStream));
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameResponseStream.cs]: Same comments from above apply here.

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameResponseStream.cs [hunk 6]
```diff
@@ -110,7 +135,7 @@ private void ValidateState()
         private enum StreamState
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameResponseStream.cs]: This enum is duplicate.

File: src/Microsoft.AspNet.Server.Kestrel/KestrelServerInformation.cs [hunk 7]
```diff
@@ -107,5 +110,23 @@ private static bool GetNoDelay(IConfiguration configuration)
...
+        {
+            var reuseStreamsString = configuration["kestrel.reuseStreams"];
+
+            if (string.IsNullOrEmpty(reuseStreamsString))
+            {
+                return false;
+            }
+
+            bool reuseStreams;
+            if (bool.TryParse(reuseStreamsString, out reuseStreams))
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/KestrelServerInformation.cs]: Does `TryParse` throw on `null`? If not you could eliminate the previous `if` block.

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                {
+                    return context.Request.Body.CopyToAsync(context.Response.Body);
+                });
+            });            
+
+            using (var app = hostBuilder.Build().Start())
+            {
+                using (var client = new HttpClient())
+                {
+
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs]: nit: remove empty line

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+using Microsoft.AspNet.Http.Features;
+using Microsoft.Extensions.Configuration;
+using Xunit;
+
+namespace Microsoft.AspNet.Server.Kestrel.FunctionalTests
+{
+    public class ReuseStreamsTests
+    {
+        [Fact]
+        public async Task ReuseStreamsOn()
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs]: I'm sorry if I'm missing something obvious here, but I don't see how this test is effectively validating that the stream are being reused.

File: test/Microsoft.AspNet.Server.KestrelTests/KestrelServerInformationTests.cs [hunk 10]
```diff
@@ -85,6 +85,38 @@ public void SetNoDelayUsingConfiguration()
             Assert.False(information.NoDelay);
         }
 
+
+        [Fact]
+        public void SetReuseStreamsUsingConfiguration()
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.KestrelTests/KestrelServerInformationTests.cs]: Please split into separate tests for `"true"` and `"false"`. Also please add tests for `null`, `string.Empty` and an invalid string (i.e. not parsable as `bool`).

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: 1) To the user of the stream it should seem like its disposed; the implementation of it being reused shouldn't be exposed. Also in the non-reuse case it stays in the DIsposed state
2) Should throw as you are in the dangerzone writing to it in the reuse case.

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs (hunk 9 above)
Comment (benaadams) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs]: Its validating that it doesn't crash when the stream is reused/not reused (on the keep-alive connection); as the initialization and re-initialization paths are different.

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs (hunk 3 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: Definitely. Concerned about the `DisposeContext` not getting called?

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs (hunk 9 above)
Comment (benaadams) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs]: Will add something to ensure resuse is and is not taking place

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs (hunk 3 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: Oh, no it can't

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs (hunk 4 above)
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: :+1:

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs (hunk 3 above)
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: I'm concerned about leaving streams in the `Closed` state.

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: If it throws, you move into the finally block and shutdown the connection and throw away the streams with it - they are only reused over the connection.

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs (hunk 2 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: Done

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs (hunk 9 above)
Comment (benaadams) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs]: Is now

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs (hunk 3 above)
Comment (cesarbs) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs]: Ok, nice.

File: test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+                        };
+                        request.Headers.Add("Connection", new string[] { "Keep-Alive" });
+                        var responseMessage = await client.SendAsync(request);
+                        var result = await responseMessage.Content.ReadAsStringAsync();
+                        Assert.Equal(content, result);
+                    }
+                }
+            }
+
+            Assert.True(streamCount == loopCount);
```
Comment (cesarbs) [inline test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/ReuseStreamsTests.cs]: Assert.Equal

**Comment by cesarbs** [comment]
LGTM, just address that `Assert.Equal` comment I've just left here.

I'll run some benchmarks on this.

Did you perform any real world tests with this change? No chance of stuff getting mixed up in the streams (i.e. something from one request spilling into the next)?

**Comment by benaadams** [comment]
> Did you perform any real world tests with this change? 

Yes, though I'm around 1.5Mrps and hit a plateau (though can get 2M+ at pipeline 512) - my guess is lock contention/spinlock on libuv thread; so working that through.

> No chance of stuff getting mixed up in the streams (i.e. something from one request spilling into the next)?

No - however, if someone held onto a stream passed `oncomplete` and after the context had been "disposed" then tried to do something with it after the next request had started - they could write to the wrong stream (the correct stream being an thrown Exception) - however that would be a bad coding error.

But... people do dumb things; which is why I put it behind a flag that's off by default. So you need to actively switch it on and make that choice.

**Comment by benaadams** [comment]
Though in effect its holding onto the socket rather than any state data; would have a similar issue if you held onto the `IFeatureCollection` and used that beyond the end of the request.

File: src/Microsoft.AspNet.Server.Kestrel/IKestrelServerInformation.cs [hunk 12]
```diff
@@ -11,6 +11,8 @@ public interface IKestrelServerInformation
 
         bool NoDelay { get; set; }
 
+        bool ReuseStreams { get; set; }
```
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Server.Kestrel/IKestrelServerInformation.cs]: Add an XML doc comment to describe what this option does please.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 13]
```diff
@@ -86,7 +87,12 @@ public Frame(ConnectionContext context)
             _localEndPoint = localEndPoint;
             _prepareRequest = prepareRequest;
             _pathBase = context.ServerAddress.PathBase;
-
+            if (ReuseStreams)
+            {
+                _requestBody = new FrameRequestStream();
+                _responseBody = new FrameResponseStream(this);
+                _duplexStream = new FrameDuplexStream(_requestBody, _responseBody);
+            }
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: nit:

File: src/Microsoft.AspNet.Server.Kestrel/IKestrelServerInformation.cs (hunk 12 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/IKestrelServerInformation.cs]: Added

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs (hunk 13 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: Added return

**Comment by halter73** [comment]
:ship:
